keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1024 @@
+-->
+
+This KEP proposes a method for provisioning volumes from cross-namespace snapshots.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+By using [volume snapshots feature](https://kubernetes.io/docs/concepts/storage/volume-snapshots/), users can provision volumes from snapshots.
+However, it only works for the `VolumeSnapshot` in the same namespace,
+therefore users can't provision a volume in one namespace from a `VolumeSnapshot` in the other namespace.
+On the other hand, as discussed in other KEP PRs (https://github.com/kubernetes/enhancements/pull/643,
+https://github.com/kubernetes/enhancements/pull/1112, and https://github.com/kubernetes/enhancements/pull/2849), there are use cases that require to share the `VolumeSnapshot` across namespaces.
+For such use cases, this KEP proposes a method for provisioning volumes from cross-namespace snapshots.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Provision of PVs from `VolumeSnapshot`s in other namespaces
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Provision of PVs from PVCs in other namespaces (Please also see [Provisioning PVs from cross-namespace PVs](#provisioning-pvs-from-cross-namespace-pvs))
+- Copy or move of `VolumeSnapshot`s to other namespaces (Please also see [Alternatives](#alternatives))
+- Clone of `VolumeSnapshotContent`s 
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Define an API to specify a cross-namespace `VolumeSnapshot` as a `DataSourceRef` of a PVC and implement a generic populator for the API.
+
+- To specify a non-standard API as a `DataSourceRef` of a PVC, [AnyVolumeDataSource feature](https://kubernetes.io/blog/2021/08/30/volume-populators-redesigned/) is used,
+- To specify a cross-namespace `VolumeSnapshot`, a new `VolumeSnapshotLink` CRD is introduced (Please also see [API](#api)),
+- To restrict only allowed `VolumeSnapshot` to be consumed from other namespaces, [`ReferencePolicy` CRD](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.ReferencePolicy) is used,
```

Comment:
Have you talked to sig-networking regarding using this CRD?  Since it is still Alpha, we should let them know we are using it and make sure the usage is supported.

Reply:
Not yet. Thank you for pointing it out.

@robscott @hbagdi @youngnick

In this KEP, we are thinking about using `ReferencePolicy` to allow cross-namespace reference of `VolumeSnapshot` for `PersistentVolumeClaim`. Could you review this KEP from `ReferencePolicy` view point?

(As I heard in the KubeCon presentation on Gateway API, there seems a plan in renaming this object, which would be one of the biggest risks).

Reply:
Thanks for checking in with us! I've created a PR that would rename ReferencePolicy to ReferenceGrant as we'd considered earlier. If/when that gets in, I'd expect the resource to be quite stable going forward.

Reply:
Thank you for sharing the current status. I will add a dependency to the PR and only use `ReferenceGrant` for this feature.

Reply:
The ReferencePolicy -> ReferenceGrant rename has gone through, so using ReferenceGrant is definitely the way to go.